src/metabase/email.clj

```diff
@@ -133,35 +167,78 @@
-                               :text        message
-                               :html        [{:type    "text/html; charset=utf-8"
-                                              :content message}])}
-                   (when-let [reply-to (email-reply-to)]
-                     {:reply-to reply-to}))))
-   (catch Throwable e
-     (prometheus/inc :metabase-email/message-errors)
-     (throw e))
-   (finally
-    (prometheus/inc :metabase-email/messages))))
+    (let [to-type (if (:bcc? email) :bcc :to)]
+      (send-email! (smtp-settings)
+                   (merge
+                    {:from    (if-let [from-name (email-from-name)]
+                                (str from-name " <" (email-from-address) ">")
+                                (email-from-address))
+                     to-type  recipients
+                     :subject subject
+                     :body    (case message-type
+                                :attachments message
+                                :text        message
+                                :html        [{:type    "text/html; charset=utf-8"
+                                               :content message}])}
+                    (when-let [reply-to (email-reply-to)]
+                      {:reply-to reply-to}))))
+    (catch Throwable e
+      (prometheus/inc :metabase-email/message-errors)
+      (when (not= :smtp-host-not-set (:cause (ex-data e)))
+        (throw e)))
+    (finally
+      (prometheus/inc :metabase-email/messages))))
+
+(defn- retry-configuration []
+  (cond-> {:max-attempts (email-retry-max-attempts)
+           :initial-interval-millis (email-retry-initial-interval)
+           :multiplier (email-retry-multiplier)
+           :randomization-factor (email-retry-randomization-factor)
+           :max-interval-millis (email-retry-max-interval-millis)}
+    (or config/is-dev? config/is-test?) (assoc :max-attempts 1)))
+
+(defn- make-retry-state
+  "Returns a notification sender wrapping [[send-password-reset-email!]] retrying
+  according to `retry-configuration`."
+  []
+  (let [retry (retry/random-exponential-backoff-retry "send-password-reset-retrying"
+                                                      (retry-configuration))]
+    {:retry retry
+     :sender (retry/decorate send-message-or-throw! retry)}))
+
+(defonce
+  ^{:private true
+    :doc "Stores the current retry state. Updated whenever the reset password
+  retry settings change. It starts with value `nil` but is set whenever the
+  settings change or when the first call with retry is made."}
+  retry-state
+  (atom nil))
+
+(defn- reconfigure-retrying [_old-value _new-value]
+  (log/info (trs "Reconfiguring reset password sender"))
+  (reset! retry-state (make-retry-state)))
+
+(defn send-email-retrying!
+  "Like [[send-password-reset-email!]] but retries sending on errors according
+  to the retry settings."
+  [& args]
+  (try
+    (when-not @retry-state
+      (compare-and-set! retry-state nil (make-retry-state)))
```

Comment:
this feels weird to me. We could likely send emails from lots of threads. Having a single global shared retry state feels quite strange.

Reply:
Ah right that was also an issue with the old pulse retry logic code, I'll take a look into creating different retry states for different events